File: hanzo-dev/zap/examples/haskell/Agent.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE LambdaCase #-}

-- | ZAP Agent Example - Haskell
--
-- Demonstrates a complete ZAP agent with tool execution in Haskell.
--
-- Usage:
--   cabal run zap-agent-haskell
--
-- Or with stack:
--   stack run

module Main where

import Control.Exception (IOException, catch)
import Control.Monad (forM)
import Data.Aeson (ToJSON(..), FromJSON(..), object, (.=), encode)
import qualified Data.ByteString.Lazy.Char8 as BL
import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map
import Data.Maybe (fromMaybe)
import Data.Text (Text)
import qualified Data.Text as T
import qualified Data.Text.IO as TIO
import System.Directory (listDirectory, doesDirectoryExist, getFileSize, getCurrentDirectory, makeAbsolute)
import System.Environment (getEnvironment)
import System.Exit (ExitCode(..))
import System.FilePath ((</>), takeFileName)
import System.Posix.Process (getProcessID)
import System.Process (readProcessWithExitCode)
import Text.Printf (printf)


-- | Approval policy enum (matches hanzo-protocol)
-- Determines when to ask for user approval before tool execution
data ApprovalPolicy
    = Untrusted    -- ^ Only auto-approve known-safe read operations
    | OnFailure    -- ^ Auto-approve, escalate on failure
    | OnRequest    -- ^ Model decides when to ask (default)
    | Never        -- ^ Never ask, return failures to model
    deriving (Show, Eq, Enum, Bounded)

instance ToJSON ApprovalPolicy where
    toJSON = \case
        Untrusted  -> "untrusted"
        OnFailure  -> "on-failure"
        OnRequest  -> "on-request"
        Never      -> "never"


-- | Sandbox policy enum (matches hanzo-protocol)
-- Determines execution restrictions for the agent
data SandboxPolicy
    = DangerFullAccess  -- ^ No restrictions (use with caution)
    | ReadOnly          -- ^ Read-only filesystem access
    | WorkspaceWrite    -- ^ Write only to cwd and specified roots
        { writableRoots       :: [FilePath]
        , networkAccess       :: Bool
        , excludeTmpdirEnvVar :: Bool
        , excludeSlashTmp     :: Bool
        , allowGitWrites      :: Bool
        }
    deriving (Show, Eq)

instance ToJSON SandboxPolicy where
    toJSON DangerFullAccess = "danger-full-access"
    toJSON ReadOnly = "read-only"
    toJSON (WorkspaceWrite roots net tmp slashTmp git) =
        object [ "mode" .= ("workspace-write" :: Text)
               , "writable_roots" .= roots
               , "network_access" .= net
               , "exclude_tmpdir_env_var" .= tmp
               , "exclude_slash_tmp" .= slashTmp
               , "allow_git_writes" .= git
               ]

-- | Default workspace-write policy
defaultWorkspaceWrite :: SandboxPolicy
defaultWorkspaceWrite = WorkspaceWrite
    { writableRoots       = []
    , networkAccess       = True
    , excludeTmpdirEnvVar = False
    , excludeSlashTmp     = False
    , allowGitWrites      = False
    }


-- | Execution context for ZAP agent
-- Contains all state needed for tool execution
data ExecutorContext = ExecutorContext
    { ctxCwd            :: FilePath
    , ctxEnv            :: Map String String
    , ctxSessionId      :: Maybe Text
    , ctxApprovalPolicy :: ApprovalPolicy
    , ctxSandboxPolicy  :: SandboxPolicy
    , ctxTimeoutMs      :: Maybe Int
    } deriving (Show, Eq)

instance ToJSON ExecutorContext where
    toJSON ExecutorContext{..} = object
        [ "cwd"             .= ctxCwd
        , "session_id"      .= ctxSessionId
        , "approval_policy" .= ctxApprovalPolicy
        , "sandbox_policy"  .= ctxSandboxPolicy
        , "timeout_ms"      .= ctxTimeoutMs
        ]

-- | Create a new ExecutorContext with sensible defaults
newExecutorContext :: FilePath -> IO ExecutorContext
newExecutorContext cwd = do
    absCwd <- makeAbsolute cwd
    env <- Map.fromList <$> getEnvironment
    pid <- getProcessID
    pure ExecutorContext
        { ctxCwd            = absCwd
        , ctxEnv            = env
        , ctxSessionId      = Just $ T.pack $ "haskell-agent-" <> show pid
        , ctxApprovalPolicy = OnRequest
        , ctxSandboxPolicy  = defaultWorkspaceWrite
        , ctxTimeoutMs      = Just 30000
        }


-- | Tool execution result
-- Uses Either for idiomatic Haskell error handling
data ToolResult a
    = ToolSuccess a
    | ToolError Text
    deriving (Show, Eq)

instance Functor ToolResult where
    fmap f (ToolSuccess a) = ToolSuccess (f a)
    fmap _ (ToolError e)   = ToolError e

instance Applicative ToolResult where
    pure = ToolSuccess
    ToolSuccess f <*> ToolSuccess a = ToolSuccess (f a)
    ToolError e   <*> _             = ToolError e
    _             <*> ToolError e   = ToolError e

instance Monad ToolResult where
    ToolSuccess a >>= f = f a
    ToolError e   >>= _ = ToolError e

-- | Convert IO exceptions to ToolResult
tryIO :: IO a -> IO (ToolResult a)
tryIO action = (ToolSuccess <$> action) `catch` handler
  where
    handler :: IOException -> IO (ToolResult a)
    handler e = pure $ ToolError $ T.pack $ show e


-- | Directory entry information
data DirEntry = DirEntry
    { entryName  :: Text
    , entryIsDir :: Bool
    , entrySize  :: Integer
    } deriving (Show, Eq)

instance ToJSON DirEntry where
    toJSON DirEntry{..} = object
        [ "name"   .= entryName
        , "is_dir" .= entryIsDir
        , "size"   .= entrySize
        ]


-- | ZAP Agent type
data ZapAgent = ZapAgent
    { agentCtx :: ExecutorContext
    } deriving (Show, Eq)

-- | Create a new ZAP agent
newZapAgent :: FilePath -> IO ZapAgent
newZapAgent cwd = ZapAgent <$> newExecutorContext cwd


-- | Available tools in this agent
listTools :: [Text]
listTools =
    [ "read_file"
    , "list_dir"
    , "git_status"
    , "git_log"
    , "exec"
    ]


-- | Read file contents
readFile' :: ExecutorContext -> FilePath -> IO (ToolResult Text)
readFile' ctx path = do
    let fullPath = ctxCwd ctx </> path
    tryIO $ TIO.readFile fullPath


-- | List directory contents
listDir :: ExecutorContext -> FilePath -> Bool -> IO (ToolResult [DirEntry])
listDir ctx path showHidden = do
    let fullPath = ctxCwd ctx </> path
    result <- tryIO $ listDirectory fullPath
    case result of
        ToolError e -> pure $ ToolError e
        ToolSuccess entries -> do
            let filtered = if showHidden
                           then entries
                           else filter (not . isHidden) entries
            ToolSuccess <$> forM filtered (mkEntry fullPath)
  where
    isHidden name = take 1 name == "."
    mkEntry dir name = do
        let entryPath = dir </> name
        isDir <- doesDirectoryExist entryPath
        size <- if isDir then pure 0 else getFileSize entryPath
        pure DirEntry
            { entryName  = T.pack name
            , entryIsDir = isDir
            , entrySize  = size
            }


-- | Get git status
gitStatus :: ExecutorContext -> IO (ToolResult Text)
gitStatus ctx = do
    (exitCode, stdout, stderr) <- readProcessWithExitCode
        "git"
        ["status", "--porcelain=v2", "--branch"]
        ""
    pure $ case exitCode of
        ExitSuccess   -> ToolSuccess $ T.pack stdout
        ExitFailure c -> ToolError $ T.pack $ "Git status failed (exit " <> show c <> "): " <> stderr


-- | Get git log
gitLog :: ExecutorContext -> Int -> IO (ToolResult Text)
gitLog ctx limit = do
    (exitCode, stdout, stderr) <- readProcessWithExitCode
        "git"
        ["log", "--format=%H|%an|%s", "-" <> show limit]
        ""
    pure $ case exitCode of
        ExitSuccess   -> ToolSuccess $ T.pack stdout
        ExitFailure c -> ToolError $ T.pack $ "Git log failed (exit " <> show c <> "): " <> stderr


-- | Execute shell command
-- Respects approval policy for security
execCommand :: ExecutorContext -> Text -> IO (ToolResult Text)
execCommand ctx command = do
    -- Check approval policy
    case ctxApprovalPolicy ctx of
        Untrusted -> pure $ ToolError "Command execution requires approval in untrusted mode"
        _ -> do
            (exitCode, stdout, stderr) <- readProcessWithExitCode
                "sh"
                ["-c", T.unpack command]
                ""
            let output = T.pack $ stdout <> stderr
            pure $ case exitCode of
                ExitSuccess   -> ToolSuccess output
                ExitFailure c -> ToolError $ "Exit code " <> T.pack (show c) <> ": " <> output


-- | Tool argument types
data ToolArgs
    = ReadFileArgs { argPath :: FilePath }
    | ListDirArgs { argPath :: FilePath, argShowHidden :: Bool }
    | GitStatusArgs
    | GitLogArgs { argLimit :: Int }
    | ExecArgs { argCommand :: Text }
    deriving (Show, Eq)


-- | Execute a tool by name with arguments
execute :: ZapAgent -> Text -> ToolArgs -> IO (ToolResult Text)
execute agent "read_file" (ReadFileArgs path) =
    readFile' (agentCtx agent) path

execute agent "list_dir" (ListDirArgs path showHidden) = do
    result <- listDir (agentCtx agent) path showHidden
    pure $ fmap (T.pack . BL.unpack . encode) result

execute agent "git_status" GitStatusArgs =
    gitStatus (agentCtx agent)

execute agent "git_log" (GitLogArgs limit) =
    gitLog (agentCtx agent) limit

execute agent "exec" (ExecArgs command) =
    execCommand (agentCtx agent) command

execute _ name _ =
    pure $ ToolError $ "Unknown tool: " <> name


-- | Helper to format tool results for display
showResult :: ToolResult Text -> String
showResult (ToolSuccess content) =
    let preview = T.take 100 content
        suffix = if T.length content > 100 then "..." else ""
    in T.unpack $ preview <> suffix
showResult (ToolError e) = "Error: " <> T.unpack e


-- | Main entry point
main :: IO ()
main = do
    putStrLn "ZAP Haskell Agent Example"
    putStrLn "========================="
    putStrLn ""

    -- Create agent
    agent <- newZapAgent "."

    -- Display context info
    putStrLn $ "Session: " <> maybe "none" T.unpack (ctxSessionId $ agentCtx agent)
    putStrLn $ "Working directory: " <> ctxCwd (agentCtx agent)
    putStrLn ""

    -- List available tools
    putStrLn $ "Available tools (" <> show (length listTools) <> "):"
    mapM_ (\t -> putStrLn $ "  - " <> T.unpack t) listTools
    putStrLn ""

    -- Example 1: Read a file
    putStrLn "Example 1: Read file"
    result1 <- execute agent "read_file" (ReadFileArgs "Cargo.toml")
    case result1 of
        ToolSuccess content ->
            printf "  Result: %d bytes read\n" (T.length content)
        ToolError e ->
            putStrLn $ "  Error: " <> T.unpack e
    putStrLn ""

    -- Example 2: List directory
    putStrLn "Example 2: List directory"
    result2 <- execute agent "list_dir" (ListDirArgs "." False)
    putStrLn $ "  Result: " <> showResult result2
    putStrLn ""

    -- Example 3: Git status
    putStrLn "Example 3: Git status"
    result3 <- execute agent "git_status" GitStatusArgs
    putStrLn $ "  Result: " <> showResult result3
    putStrLn ""

    -- Example 4: Permission check (demonstrate approval policy)
    putStrLn "Example 4: Permission check"
    let ctx = agentCtx agent
    putStrLn $ "  Approval policy: " <> show (ctxApprovalPolicy ctx)
    putStrLn $ "  Sandbox policy: " <> case ctxSandboxPolicy ctx of
        DangerFullAccess -> "danger-full-access"
        ReadOnly -> "read-only"
        WorkspaceWrite{} -> "workspace-write"

    -- Example 5: Execute a command
    putStrLn ""
    putStrLn "Example 5: Execute command"
    result5 <- execute agent "exec" (ExecArgs "echo 'Hello from ZAP Haskell agent!'")
    putStrLn $ "  Result: " <> showResult result5

File: hanzo-dev/zap/examples/go/agent.go
// ZAP Agent Example - Go
//
// Demonstrates a complete ZAP agent with tool execution in Go.
//
// Usage:
//   go run agent.go

package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

// AskForApproval determines when to ask for user approval
type AskForApproval string

const (
	ApprovalUntrusted AskForApproval = "untrusted"
	ApprovalOnFailure AskForApproval = "on-failure"
	ApprovalOnRequest AskForApproval = "on-request"
	ApprovalNever     AskForApproval = "never"
)

// SandboxMode determines execution restrictions
type SandboxMode string

const (
	SandboxDangerFullAccess SandboxMode = "danger-full-access"
	SandboxReadOnly         SandboxMode = "read-only"
	SandboxWorkspaceWrite   SandboxMode = "workspace-write"
)

// ExecutorContext contains execution state
type ExecutorContext struct {
	Cwd            string
	Env            map[string]string
	SessionID      string
	ApprovalPolicy AskForApproval
	SandboxMode    SandboxMode
	WritableRoots  []string
	NetworkAccess  bool
	TimeoutMs      int64
}

// ToolResult represents a tool execution result
type ToolResult struct {
	Content interface{} `json:"content"`
	Error   string      `json:"error,omitempty"`
}

// ZapAgent is a ZAP-compatible agent
type ZapAgent struct {
	ctx ExecutorContext
}

// NewZapAgent creates a new agent
func NewZapAgent(cwd string) *ZapAgent {
	absPath, _ := filepath.Abs(cwd)
	return &ZapAgent{
		ctx: ExecutorContext{
			Cwd:            absPath,
			Env:            envToMap(),
			SessionID:      fmt.Sprintf("go-agent-%d", os.Getpid()),
			ApprovalPolicy: ApprovalOnRequest,
			SandboxMode:    SandboxWorkspaceWrite,
			WritableRoots:  []string{},
			NetworkAccess:  true,
			TimeoutMs:      30000,
		},
	}
}

func envToMap() map[string]string {
	env := make(map[string]string)
	for _, e := range os.Environ() {
		parts := strings.SplitN(e, "=", 2)
		if len(parts) == 2 {
			env[parts[0]] = parts[1]
		}
	}
	return env
}

// Execute runs a tool by name
func (a *ZapAgent) Execute(name string, args map[string]interface{}) ToolResult {
	switch name {
	case "read_file":
		return a.readFile(args["path"].(string))
	case "list_dir":
		return a.listDir(args["path"].(string))
	case "git_status":
		return a.gitStatus()
	case "git_log":
		limit := 10
		if l, ok := args["limit"].(float64); ok {
			limit = int(l)
		}
		return a.gitLog(limit)
	case "exec":
		return a.exec(args["command"].(string))
	default:
		return ToolResult{Error: fmt.Sprintf("Unknown tool: %s", name)}
	}
}

func (a *ZapAgent) readFile(path string) ToolResult {
	fullPath := filepath.Join(a.ctx.Cwd, path)
	content, err := ioutil.ReadFile(fullPath)
	if err != nil {
		return ToolResult{Error: err.Error()}
	}
	return ToolResult{Content: string(content)}
}

func (a *ZapAgent) listDir(path string) ToolResult {
	fullPath := filepath.Join(a.ctx.Cwd, path)
	entries, err := ioutil.ReadDir(fullPath)
	if err != nil {
		return ToolResult{Error: err.Error()}
	}

	result := make([]map[string]interface{}, 0, len(entries))
	for _, e := range entries {
		result = append(result, map[string]interface{}{
			"name":  e.Name(),
			"isDir": e.IsDir(),
			"size":  e.Size(),
		})
	}
	return ToolResult{Content: result}
}

func (a *ZapAgent) gitStatus() ToolResult {
	cmd := exec.Command("git", "status", "--porcelain=v2", "--branch")
	cmd.Dir = a.ctx.Cwd
	output, err := cmd.Output()
	if err != nil {
		return ToolResult{Error: err.Error()}
	}
	return ToolResult{Content: string(output)}
}

func (a *ZapAgent) gitLog(limit int) ToolResult {
	cmd := exec.Command("git", "log", fmt.Sprintf("--format=%%H|%%an|%%s"), fmt.Sprintf("-%d", limit))
	cmd.Dir = a.ctx.Cwd
	output, err := cmd.Output()
	if err != nil {
		return ToolResult{Error: err.Error()}
	}
	return ToolResult{Content: string(output)}
}

func (a *ZapAgent) exec(command string) ToolResult {
	// Check approval policy
	if a.ctx.ApprovalPolicy == ApprovalUntrusted {
		return ToolResult{Error: "Command execution requires approval in untrusted mode"}
	}

	cmd := exec.Command("sh", "-c", command)
	cmd.Dir = a.ctx.Cwd
	output, err := cmd.CombinedOutput()
	if err != nil {
		return ToolResult{Content: string(output), Error: err.Error()}
	}
	return ToolResult{Content: string(output)}
}

// ListTools returns available tools
func (a *ZapAgent) ListTools() []string {
	return []string{
		"read_file",
		"list_dir",
		"git_status",
		"git_log",
		"exec",
	}
}

func main() {
	fmt.Println("ZAP Go Agent Example")
	fmt.Println("====================")
	fmt.Println()

	agent := NewZapAgent(".")

	// List available tools
	tools := agent.ListTools()
	fmt.Printf("Available tools (%d):\n", len(tools))
	for _, t := range tools {
		fmt.Printf("  - %s\n", t)
	}
	fmt.Println()

	// Example 1: Read a file
	fmt.Println("Example 1: Read file")
	result := agent.Execute("read_file", map[string]interface{}{"path": "Cargo.toml"})
	if result.Error != "" {
		fmt.Printf("  Error: %s\n", result.Error)
	} else {
		content := result.Content.(string)
		fmt.Printf("  Result: %d bytes read\n", len(content))
	}
	fmt.Println()

	// Example 2: List directory
	fmt.Println("Example 2: List directory")
	result = agent.Execute("list_dir", map[string]interface{}{"path": "."})
	if result.Error != "" {
		fmt.Printf("  Error: %s\n", result.Error)
	} else {
		jsonBytes, _ := json.Marshal(result.Content)
		preview := string(jsonBytes)
		if len(preview) > 100 {
			preview = preview[:100] + "..."
		}
		fmt.Printf("  Result: %s\n", preview)
	}
	fmt.Println()

	// Example 3: Git status
	fmt.Println("Example 3: Git status")
	result = agent.Execute("git_status", map[string]interface{}{})
	if result.Error != "" {
		fmt.Printf("  Error: %s\n", result.Error)
	} else {
		content := result.Content.(string)
		preview := content
		if len(preview) > 100 {
			preview = preview[:100] + "..."
		}
		fmt.Printf("  Result: %s\n", preview)
	}
}
